File: utils/utils-base/src/com/xenoage/utils/font/FontStyle.java
package com.xenoage.utils.font;

import com.xenoage.utils.annotations.Const;

/**
 * Enumeration of different font styles like bold or italic.
 * 
 * @author Andreas Wenger
 */
@Const public class FontStyle {

	public static final int Bold = 1 << 0;
	public static final int Italic = 1 << 1;
	public static final int Underline = 1 << 2;
	public static final int Strikethrough = 1 << 3;

	public static final FontStyle normal = new FontStyle(0);

	private final int style;

	
	public static FontStyle fontStyle(int... flags) {
		int style = 0;
		for (int flag : flags)
			style |= flag;
		if (style == 0)
			return normal;
		else
			return new FontStyle(style);
	}

	public FontStyle(int style) {
		this.style = style;
	}

	/**
	 * Sets the given flag, like Bold, to true.
	 */
	public FontStyle with(int flag) {
		return with(flag, true);
	}

	/**
	 * Sets the given flag, like Bold, to the given value.
	 */
	public FontStyle with(int flag, boolean value) {
		int style = this.style;
		if (value)
			style |= flag;
		else
			style &= ~flag;
		return new FontStyle(style);
	}

	/**
	 * Returns true, when the given flag is set.
	 */
	public boolean isSet(int flag) {
		return (style & flag) > 0;
	}
	
	@Override public int hashCode() {
		return style;
	}

	@Override public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		FontStyle other = (FontStyle) obj;
		if (style != other.style)
			return false;
		return true;
	}

	@Override public String toString() {
		return (isSet(Bold) ? "B" : "b") + (isSet(Italic) ? "I" : "i") +
			(isSet(Underline) ? "U" : "u") + (isSet(Strikethrough) ? "S" : "s");
	}

}

File: utils-common/src/com/xenoage/utils/font/FontStyle.kt
package com.xenoage.utils.font

import com.xenoage.utils.Cache
import com.xenoage.utils.annotations.Optimized
import com.xenoage.utils.annotations.Reason
import com.xenoage.utils.annotations.Reason.MemorySaving

/**
 * Enumeration of different font styles like bold or italic.
 */
data class FontStyle private constructor(
		val bold: Boolean,
		val italic: Boolean,
		val underline: Boolean,
		val strikethrough: Boolean
) {

	override fun toString(): String {
		return (if (bold) "B" else "b") + (if (italic) "I" else "i") +
				(if (underline) "U" else "u") + if (strikethrough) "S" else "s"
	}

	companion object {

		val normal = FontStyle(false, false, false, false)

		var cache = Cache<Int, FontStyle>(16) //16 possibilities

		@Optimized(MemorySaving)
		operator fun invoke(bold: Boolean = false, italic: Boolean = false,
		                    underline: Boolean = false, strikethrough: Boolean = false): FontStyle {
			val id = if (bold) 8 else 0 + if (italic) 4 else 0 + if (underline) 2 else 0 + if (strikethrough) 1 else 0
			return cache[id, { FontStyle(bold, italic, underline, strikethrough) }]
		}

	}

}
